### PR TITLE
Other/error handling/provider/mongodb

### PR DESCRIPTION
**Pull request by RiddhiBagadiaa**
# Description
Making go errors more structured

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have fixed any merge conflicts

### CODE REVIEW

File: provider/mongodb.go [hunk 1]
```diff
@@ -40,7 +41,7 @@ type mongoDBOnlineTable struct {
 func mongoOnlineStoreFactory(serialized pc.SerializedConfig) (Provider, error) {
 	mongoConfig := &pc.MongoDBConfig{}
 	if err := mongoConfig.Deserialize(serialized); err != nil {
-		return nil, err
+		return nil, fferr.NewInternalError(err)
```
Comment (sdreyer) [inline provider/mongodb.go:44]: Dont need to wrap this

File: provider/mongodb.go [hunk 2]
```diff
@@ -50,16 +51,16 @@ func NewMongoDBOnlineStore(config *pc.MongoDBConfig) (*mongoDBOnlineStore, error
 	uri := fmt.Sprintf("mongodb://%s:%s@%s:%s/?ssl=true&replicaSet=globaldb&retrywrites=false&maxIdleTimeMS=120000", config.Username, config.Password, config.Host, config.Port)
 	client, err := mongo.Connect(context.TODO(), options.Client().ApplyURI(uri))
 	if err != nil {
-		return nil, fmt.Errorf("could not connect to mongodb: %w", err)
+		return nil, fferr.NewConnectionError("mongodb", err)
 	}
 	cur, err := client.Database(config.Database).ListCollections(context.TODO(), bson.D{{"name", "featureform__metadata"}})
 	if err != nil {
-		return nil, fmt.Errorf("could not create check if metadata exists: %w", err)
+		return nil, fferr.NewInternalError(err)
```
Comment (sdreyer) [inline provider/mongodb.go:58]: probably better as an execution error

File: provider/mongodb.go [hunk 3]
```diff
@@ -120,14 +121,14 @@ func (store *mongoDBOnlineStore) CreateTable(feature, variant string, valueType
...
-		return nil, fmt.Errorf("could not insert metadata table name: %w", err)
+		return nil, fferr.NewExecutionError("mongodb", feature, variant, "", err)
 	}
 
 	command := bson.D{{"customAction", "CreateCollection"}, {"collection", tableName}, {"autoScaleSettings", bson.D{{"maxThroughput", store.tableThroughput}}}}
 	var cmdResult interface{}
 	err = store.client.Database(store.database).RunCommand(context.TODO(), command).Decode(&cmdResult)
 	if err != nil {
-		return nil, fmt.Errorf("could not set table throughput: %s, %w", tableName, err)
+		return nil, fferr.NewExecutionError("mongodb", feature, variant, "", err)
```
Comment (sdreyer) [inline provider/mongodb.go:131]: type field should be a feature

File: provider/mongodb.go [hunk 4]
```diff
@@ -145,12 +146,12 @@ func (store *mongoDBOnlineStore) GetTable(feature, variant string) (OnlineStoreT
 	tableName := store.GetTableName(feature, variant)
 	cur, err := store.client.Database(store.database).ListCollections(context.TODO(), bson.D{{"name", tableName}})
 	if err != nil {
-		return nil, fmt.Errorf("could not create check if metadata exists: %w", err)
+		return nil, fferr.NewExecutionError("mongodb", feature, variant, "", err)
```
Comment (sdreyer) [inline provider/mongodb.go:149]: type param should be feature

Comment (sdreyer) [reply provider/mongodb.go:149]: same for all below

File: provider/mongodb.go [hunk 5]
```diff
@@ -222,7 +223,7 @@ func (table mongoDBOnlineTable) Get(entity string) (interface{}, error) {
 			fmt.Printf("could not get table value: %s: %s: %s", table.name, entity, err.Error())
 			return nil, &EntityNotFound{entity}
 		}
-		return nil, fmt.Errorf("could not get table value: %s: %s: %w", table.name, entity, err)
+		return nil, fferr.NewInternalError(err)
```
Comment (sdreyer) [inline provider/mongodb.go:226]: execution error

File: provider/mongodb.go [hunk 6]
```diff
@@ -146,7 +146,7 @@ func (store *mongoDBOnlineStore) GetTable(feature, variant string) (OnlineStoreT
 	tableName := store.GetTableName(feature, variant)
 	cur, err := store.client.Database(store.database).ListCollections(context.TODO(), bson.D{{"name", tableName}})
 	if err != nil {
-		return nil, fferr.NewExecutionError("mongodb", feature, variant, "", err)
+		return nil, fferr.NewExecutionError("mongodb", feature, variant, "feature", err)
```
Comment (sdreyer) [inline provider/mongodb.go:149]: @epps is this FEATURE_VARIANT?

File: provider/mongodb.go [hunk 7]
```diff
@@ -203,7 +204,7 @@ func (table mongoDBOnlineTable) Set(entity string, value interface{}) error {
 			},
 		)
 	if err != nil {
-		return fmt.Errorf("could not set values: (entity: %s, value: %v): %w", entity, value, err)
+		return fferr.NewInternalError(err)
```
Comment (sdreyer) [inline provider/mongodb.go:207]: Should be execution error

**Review by epps — APPROVED** [review]
🚀

**Pull request merged**